src/coprocessor/statistics/analyze.rs

```diff
@@ -271,6 +291,232 @@ impl<S: Snapshot> RequestHandler for AnalyzeContext<S> {
+                                        val
+                                    } else {
+                                        // Only if the `decoded_val` is Datum::Null, `decoded_val` is a Ok(None).
+                                        // So it is safe the unwrap the Ok value.
+                                        let decoded_sorted_val = TT::sort_key(&decoded_val.as_string()?.unwrap().into_owned())?;
+                                        encode_value(&mut EvalContext::default(), &[Datum::Bytes(decoded_sorted_val)])?
+                                    }
+                                }
+                            }
+                        };
+                        column_vals.push(sorted_val);
+                        continue;
+                    }
+                    column_vals.push(val);
+                }
+                collector.count += 1;
+                collector.collect_column_group(&column_vals, &self.column_groups);
+                collector.collect_column(column_vals);
+            }
+        }
+        Ok(AnalyzeSamplingResult::new(collector))
+    }
+}
+
+#[derive(Clone)]
+struct RowSampleCollector {
+    samples: BinaryHeap<Reverse<(i64, Vec<Vec<u8>>)>>,
+    null_count: Vec<i64>,
+    count: u64,
+    max_sample_size: usize,
+    fm_sketches: Vec<FmSketch>,
+    rng: StdRng,
+    total_sizes: Vec<i64>,
+    row_buf: Vec<u8>,
+}
+
+impl Default for RowSampleCollector {
+    fn default() -> Self {
+        RowSampleCollector {
+            samples: BinaryHeap::new(),
+            null_count: vec![],
+            count: 0,
+            max_sample_size: 0,
+            fm_sketches: vec![],
+            rng: StdRng::from_entropy(),
+            total_sizes: vec![],
+            row_buf: Vec::new(),
+        }
+    }
+}
+
+impl RowSampleCollector {
+    fn new(
+        max_sample_size: usize,
+        max_fm_sketch_size: usize,
+        col_and_group_len: usize,
+    ) -> RowSampleCollector {
+        RowSampleCollector {
+            samples: BinaryHeap::new(),
+            null_count: vec![0; col_and_group_len],
+            count: 0,
+            max_sample_size,
+            fm_sketches: vec![FmSketch::new(max_fm_sketch_size); col_and_group_len],
+            rng: StdRng::from_entropy(),
+            total_sizes: vec![0; col_and_group_len],
+            row_buf: Vec::new(),
+        }
+    }
+
+    pub fn collect_column_group(
+        &mut self,
+        columns_val: &[Vec<u8>],
+        column_groups: &[tipb::AnalyzeColumnGroup],
+    ) {
+        self.row_buf.clear();
+        let col_len = columns_val.len();
+        for i in 0..column_groups.len() {
+            let offsets = column_groups[i].get_column_offsets();
+            let mut has_null = true;
+            for j in 0..offsets.len() {
+                if columns_val[j][0] != NIL_FLAG {
+                    has_null = false
+                }
+                self.total_sizes[col_len + i] += columns_val[i].len() as i64
```

Comment:
The Golang version is `s.TotalSizes[colLen+i] += int64(len(cols[c].GetBytes())) - 1`. Why not minus one here?

Reply:
There's a tiny difference between the two version.
For golang version, it called the `DecodeColumnValue` which called `DecodeOne` first then called the `Unflatten`. In the `DecodeOne` you can see that the first byte in the slice means the `type` information.
In rust version, it directly called a `unflatten` in the `decode_col_value` and judge the `type` by the given `ColumnInfo` so the byte slice passed here contains no `type` information. Thus we don't minus one in the rust version.